File: ios/GuardApp/Core/Bluetooth/RollingIdentifier.swift
import CryptoKit
import Foundation

struct RollingIdentifier: Equatable {
    let value: String
    let timeBucket: Int64
}

final class RollingIdentifierGenerator {
    private let bucketSeconds: Int64
    private let outputBytes: Int

    init(bucketSeconds: Int64 = 300, outputBytes: Int = 12) {
        precondition(bucketSeconds > 0)
        precondition(outputBytes > 0 && outputBytes <= 32)
        self.bucketSeconds = bucketSeconds
        self.outputBytes = outputBytes
    }

    func generate(groupSecret: Data, deviceId: String, date: Date) -> RollingIdentifier {
        let bucket = Int64(date.timeIntervalSince1970) / bucketSeconds
        let message = "\(deviceId):\(bucket)"
        let key = SymmetricKey(data: groupSecret)
        let signature = HMAC<SHA256>.authenticationCode(for: Data(message.utf8), using: key)
        let value = Data(signature).prefix(outputBytes).map { String(format: "%02x", $0) }.joined()
        return RollingIdentifier(value: value, timeBucket: bucket)
    }

    func candidates(groupSecret: Data, deviceIds: [String], date: Date, bucketDrift: Int64 = 1) -> [String: String] {
        var result: [String: String] = [:]
        let currentBucket = Int64(date.timeIntervalSince1970) / bucketSeconds
        for deviceId in deviceIds {
            for bucket in (currentBucket - bucketDrift)...(currentBucket + bucketDrift) {
                let message = "\(deviceId):\(bucket)"
                let key = SymmetricKey(data: groupSecret)
                let signature = HMAC<SHA256>.authenticationCode(for: Data(message.utf8), using: key)
                let value = Data(signature).prefix(outputBytes).map { String(format: "%02x", $0) }.joined()
                result[value] = deviceId
            }
        }
        return result
    }
}

File: android/app/src/main/java/com/deviceguard/core/bluetooth/RollingIdentifier.kt
package com.deviceguard.core.bluetooth

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

data class RollingIdentifier(
    val value: String,
    val timeBucket: Long
)

class RollingIdentifierGenerator(
    private val bucketSeconds: Long = 300,
    private val outputBytes: Int = 12
) {
    init {
        require(bucketSeconds > 0) { "bucketSeconds must be positive" }
        require(outputBytes in 1..32) { "outputBytes must be 1..32" }
    }

    fun generate(groupSecret: ByteArray, deviceId: String, epochSeconds: Long): RollingIdentifier {
        val bucket = epochSeconds / bucketSeconds
        return RollingIdentifier(
            value = hmacHex(groupSecret, "$deviceId:$bucket").take(outputBytes * 2),
            timeBucket = bucket
        )
    }

    fun candidates(
        groupSecret: ByteArray,
        deviceIds: List<String>,
        epochSeconds: Long,
        bucketDrift: Long = 1
    ): Map<String, String> {
        val currentBucket = epochSeconds / bucketSeconds
        return buildMap {
            for (deviceId in deviceIds) {
                for (bucket in (currentBucket - bucketDrift)..(currentBucket + bucketDrift)) {
                    val value = hmacHex(groupSecret, "$deviceId:$bucket").take(outputBytes * 2)
                    put(value, deviceId)
                }
            }
        }
    }

    private fun hmacHex(secret: ByteArray, payload: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        return mac.doFinal(payload.toByteArray(Charsets.UTF_8)).joinToString("") { "%02x".format(it) }
    }
}
